plan/logical_plan_builder.go

```diff
@@ -668,31 +668,76 @@ func (b *planBuilder) buildProjection4Union(u *LogicalUnionAll) {
 }
 
 func (b *planBuilder) buildUnion(union *ast.UnionStmt) LogicalPlan {
+	unionDistinctStmts, unionAllStmts := classifyByDistinct(union.SelectList.Selects)
+
+	unionDistinctPlan := b.buildSubUnion(unionDistinctStmts, true, nil)
+	if b.err != nil {
+		return nil
+	}
+	unionAllPlan := b.buildSubUnion(unionAllStmts, false, unionDistinctPlan)
+	if b.err != nil {
+		return nil
+	}
+
+	var unionPlan LogicalPlan
+	if unionAllPlan != nil {
+		unionPlan = unionAllPlan
+	} else if unionDistinctPlan != nil {
+		unionPlan = unionDistinctPlan
+	} else {
+		return nil
+	}
+
+	if union.OrderBy != nil {
+		unionPlan = b.buildSort(unionPlan, union.OrderBy.Items, nil)
+	}
+	if union.Limit != nil {
+		unionPlan = b.buildLimit(unionPlan, union.Limit)
+	}
+	return unionPlan
+}
+
+// classifyByDistinct to divide selects into union-all and union-distinct slices.
+// https://dev.mysql.com/doc/refman/5.7/en/union.html
+func classifyByDistinct(selects []*ast.SelectStmt) (unionDistinct []*ast.SelectStmt, unionAll []*ast.SelectStmt) {
+	firstUnionAllIdx := len(selects)
+	for ; firstUnionAllIdx > 0; firstUnionAllIdx-- {
+		if selects[firstUnionAllIdx-1].IsAfterUnionDistinct {
+			break
+		}
+	}
+	return selects[:firstUnionAllIdx], selects[firstUnionAllIdx:]
+}
+
+func (b *planBuilder) buildSubUnion(selects []*ast.SelectStmt, needDistinct bool, subPlan LogicalPlan) LogicalPlan {
+	if selects == nil || len(selects) == 0 {
+		return nil
+	}
 	u := LogicalUnionAll{}.init(b.ctx)
-	u.children = make([]LogicalPlan, len(union.SelectList.Selects))
-	for i, sel := range union.SelectList.Selects {
-		u.children[i] = b.buildSelect(sel)
+	childLen := len(selects)
+	if subPlan != nil {
+		childLen = childLen + 1
+	}
+	u.children = make([]LogicalPlan, childLen)
+	i := 0
+	for ; i < len(selects); i++ {
```

Comment:
for _, select := range selectes {
    u.children = append(u.children, b.buildSelect(select))
}